aws-lc-rs/src/cipher/streaming.rs

```diff
@@ -0,0 +1,527 @@
+        })
+    }
+
+    /// Encrypt the input and return the output.
+    /// # Errors
+    /// Returns an error if the output buffer is too small.
+    pub fn update<'a>(
+        &mut self,
+        input: &[u8],
+        output: &'a mut [u8],
+    ) -> Result<&'a [u8], Unspecified> {
+        if output.len() < (input.len() + self.algorithm.block_len) {
+            return Err(Unspecified);
+        }
+
+        let mut outlen: i32 = output.len().try_into()?;
+        let inlen: i32 = input.len().try_into()?;
+        if 1 != unsafe {
+            EVP_EncryptUpdate(
+                self.cipher_ctx.as_mut_ptr(),
+                output.as_mut_ptr(),
+                &mut outlen,
+                input.as_ptr(),
+                inlen,
+            )
+        } {
+            return Err(Unspecified);
+        }
+        let outlen: usize = outlen.try_into()?;
+        Ok(&output[0..outlen])
+    }
+
+    /// Finish the encryption and return the output.
+    /// # Errors
+    /// Returns an error if the output buffer is too small.
+    pub fn finish(self, output: &mut [u8]) -> Result<(DecryptionContext, &[u8]), Unspecified> {
+        if output.len() < self.algorithm.block_len {
+            return Err(Unspecified);
+        }
+        let mut outlen: i32 = output.len().try_into()?;
+        if 1 != unsafe {
+            EVP_EncryptFinal_ex(
+                self.cipher_ctx.as_mut_ptr(),
+                output.as_mut_ptr(),
+                &mut outlen,
+            )
+        } {
+            return Err(Unspecified);
+        }
+        let outlen: usize = outlen.try_into()?;
+        Ok((self.context.into(), &output[0..outlen]))
+    }
+
+    /// Returns the cipher operating mode.
+    #[must_use]
+    pub fn mode(&self) -> OperatingMode {
+        self.mode
+    }
+
+    /// Returns the cipher algorithm
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        self.algorithm
+    }
+
+    /// CTR cipher mode
+    /// # Errors
+    /// If the key is not valid for the cipher algorithm
+    pub fn ctr(key: UnboundCipherKey) -> Result<Self, Unspecified> {
+        let context = key.algorithm().new_encryption_context(OperatingMode::CTR)?;
+        Self::less_safe_ctr(key, context)
+    }
+
+    /// CTR cipher mode
+    /// # Errors
+    /// If the key is not valid for the cipher algorithm
+    pub fn less_safe_ctr(
+        key: UnboundCipherKey,
+        context: EncryptionContext,
+    ) -> Result<Self, Unspecified> {
+        Self::new(key, OperatingMode::CTR, context)
+    }
+
+    /// CBC cipher mode
+    /// # Errors
+    /// If the key is not valid for the cipher algorithm
+    pub fn cbc_pkcs7(key: UnboundCipherKey) -> Result<Self, Unspecified> {
+        let context = key.algorithm().new_encryption_context(OperatingMode::CBC)?;
+        Self::less_safe_cbc_pkcs7(key, context)
+    }
+
+    /// CBC cipher mode
+    /// # Errors
+    /// If the key is not valid for the cipher algorithm
+    pub fn less_safe_cbc_pkcs7(
```

Comment:
why is this less safe? Most of these methods could use additional documentation with more details